### ISSUE TITLE
Feature request: publish multi-arch docker images, for Raspberry Pi, etc.

### ISSUE THREAD

**Issue by idcrook** [alexellis/inlets#60]
## Expected Behaviour

expect/want this (on Raspberry Pi):
```shell
$ docker run -i alexellis2/inlets:2.0.0
...
Error: unknown shorthand flag: 'e' in -elp
```
with the rest of the help message, which exhibits `inlets` running successfully. This result was achieved on amd64 Linux.

_Aside_: `Dockerfile` should probably be patch thusly:
```diff
-CMD ["-help"]
+CMD ["--help"]
```
## Current Behaviour

Anyway, instead there is an arch error on raspberry pi (Raspbian, 3B+, armv7l) when using the same docker run command:
```shell
$ docker run -i alexellis2/inlets:2.0.0
```
`standard_init_linux.go:207: exec user process caused "exec format error"`

This is because the only containers (currently) on Docker Hub for `inlets` are amd64, and none of the ARM exec formats. See <https://hub.docker.com/r/alexellis2/inlets/tags>

## Possible Solution

I do not have much direct experience with multi-arch container publishing. However, I noticed [miniflux](https://miniflux.app), also a golang project, has a multi-arch `Dockerfile` with supporting `Makefile` and publishes various arch containers to Docker Hub for each release.
 - <https://github.com/miniflux/miniflux/blob/master/Dockerfile>
 - <https://github.com/miniflux/miniflux/blob/master/Makefile>
 - <https://hub.docker.com/r/miniflux/miniflux/tags>

So those might provide useful methods to draw upon? Of course, if that route was chosen, the example kubernetes deployment would need to follow: so, e.g. would need to specify `image: alexellis2/inlets:arm32v6-2.0.0`, instead of just  `image: alexellis2/inlets:2.0.0`

## Context

Along with standing up kubernetes and OpenFaaS on k8s on Ras Pi's, having native inlets support "OOTB" on the Ras Pi platform through docker/k8s would also be useful.

**Comment by idcrook** [comment]
Just noticed that the docker targets, and especially the `docker-manifest` target, in the miniflux [Makefile#L120-L150](https://github.com/miniflux/miniflux/blob/0dff43233791ae7949ea6f9a50326bbb3f3769e5/Makefile#L120-L150) show the recipe to of how annotate to images to get multi-arch "magic"‼️  That is, the image **:tag** can be just the version (without explicit arch).

I could take a pass at a PR for this issue unless someone else would prefer to.

**Comment by alexellis** [comment]
I've made some changes and how we have:

```
alexellis2/inlets:2.0.2-arm64
alexellis2/inlets:2.0.2-armhf
alexellis2/inlets:2.0.2
```

The arm* builds were done on-device manually. It shouldn't be too hard to install tooling into travis to force a build into different architectures. The primary issue is that we need to add ca-certs and that has to be run on device or via qemu. qemu needs to be patched in the kernel of the build machine to work correctly.

**Comment by alexellis** [comment]
I also updated the Dockerfile, thanks for finding that.

**Comment by idcrook** [comment]
Thanks. The following is successful (shows help message) on a Raspberry Pi (Raspbian, 3B+) 👍 
```shell
 $ docker run -i alexellis2/inlets:2.0.2-armhf
```
I think I understand what you are describing about the travis-ci. In an attempt to be helpful, I found some related questions with answers  (c. 2016) on Stack Overflow about multi-arch docker, along with referecing https://github.com/travis-ci/travis-ci/issues/3376
 - [Travis Multi-Architecture Buiilds](https://stackoverflow.com/questions/39432668/travis-multi-architecture-buiilds)
 - [Travis Build Multiple Docker Images from a Single Repo](https://stackoverflow.com/questions/39422135/travis-build-multiple-docker-images-from-a-single-repo?rq=1)

The first SO link has a "workaround" for using qemu to generate the build in travis.
The second uses a "build matrix" on travis, with multiple Dockerfile-s and [docker-flow-proxy](https://github.com/docker-flow/docker-flow-proxy)

Not sure whether either of these are best practices in 2019 for multi-arch travis-ci + docker. 

 @alexellis did you want to leave this issue open until the "release" does all the arches to docker fully automated?

**Comment by idcrook** [comment]
I was thinking about the ca-certificates issue that was raised and thought it might be possible to avoid the problem by using "pre-built" images. I have a branch [multi-arch-docker](https://github.com/idcrook/inlets/tree/multi-arch-docker) (have not yet submit PR with it) that creates a Dockerfile for each supported arch. These get built and published to the registry ahead of time, e.g. docker hub [dpcrook/alpine-inlets-arm32v6](https://cloud.docker.com/repository/docker/dpcrook/alpine-inlets-arm32v6/tags)

Then in a multi-arch Dockerfile on travis, the cross-built binaries could be copied in a matching build phase. For the release phase, Dockerfile might look something like:

```Dockerfile
FROM dpcrook/alpine-inlets-amd64:1.0 AS release-img-linux-amd64
COPY --from=build /usr/bin/inlets /usr/bin/
ENTRYPOINT ["inlets"]
CMD ["--help"]

FROM dpcrook/alpine-inlets-arm64v8:1.0 AS release-img-linux-arm64
COPY --from=build /usr/bin/inlets-arm64 /usr/bin/inlets
ENTRYPOINT ["inlets"]
CMD ["--help"]

FROM dpcrook/alpine-inlets-arm64v8:1.0 AS release-img-linux-armhf
COPY --from=build /usr/bin/inlets-armhf /usr/bin/inlets
ENTRYPOINT ["inlets"]
CMD ["--help"]
```
This is possible since COPY and ENTRYPOINT in Dockerfiles can execute on any host platform.

Each of those images has the other releases steps already present (e.g. [Dockerfile.armhf](https://github.com/idcrook/inlets/blob/multi-arch-docker/hack/Dockerfile.armhf)) The proper binary would get copied, then the matching docker build phase could be run in travis and resulting image pushed independently...

If the assumption is that the alpine base images and `ca-certificates` do not change often, then something like this might be adequate? If this route is used, there is a direct coupling between the release and the image it is derived from, so those parent images would need to stay around as long as their derived images.

**Comment by idcrook** [comment]
In my fork https://github.com/idcrook/inlets/commit/ec8ee5973fe95f990f883d4d54b3908de5cb1c1b I have successfully demonstrated multi-arch dockerfile using the BuildKit approach. It includes built-in publishing support (refer to my docker hub [test repo](https://cloud.docker.com/repository/docker/dpcrook/inlets/tags))

These published images with pulled from docker hub,
- successfully ran `docker run -i dpcrook/inlets:test-armhf` on a raspberry pi
- successfully ran `docker run -i dpcrook/inlets:test-arm64` on an aarch64
- successfully ran on amd64 host

```bash
$ sudo docker run -i dpcrook/inlets:test-arm64 --version
Version: 2.0.3-alpha
Git Commit: b9daa7d
$ uname -m
aarch64
```

Though I used them for proof-of-concept and debugging, I don't think the intermediate images to derive from are needed. It should be possible to consolidate everything into a single multi-arch Dockerfile.

Let me know how you'd like to proceed.
